src/build/opBuildAPIClient.ts

```diff
@@ -0,0 +1,87 @@
+import * as request from 'request';
+import * as querystring from 'querystring';
+import { EnvironmentController } from '../common/EnvironmentController';
+import { EventStream } from '../common/EventStream';
+import { extensionConfig } from '../shared';
+import { APICallFailed, APICallStarted, CredentialExpired } from '../common/loggingEvents';
+
+const OP_BUILD_USER_TOKEN_HEADER_NAME = 'X-OP-BuildUserToken';
+
+export class OPBuildAPIClient {
+    constructor(private environmentController: EnvironmentController) { }
+
+    public async getOriginalRepositoryUrl(gitRepoUrl: string, opBuildUserToken: string, eventStream: EventStream): Promise<string> {
+        const requestUrl = `${this.APIBaseUrl}/v2/Repositories/OriginalRepositoryUrl`
+            + `?${querystring.stringify({
+                gitRepoUrl
+            })}`;
+
+        return this.sendRequest('GetOriginalRepositoryUrl', requestUrl, 'GET', opBuildUserToken, eventStream)
+            .then((response: any) => {
+                if (this.isResultEmpty(response)) {
+                    throw new Error(`Cannot the get original repository URL for current repository(The repository is not provisioned in Docs).`);
+                }
+                return response.toString();
+            }, (code: any) => {
+                throw new Error(`Cannot the get original repository URL for current repository(${code})`);
+            });
+    }
+
+    private get APIBaseUrl() {
+        return extensionConfig.OPBuildAPIEndPoint[this.environmentController.env];
+    }
+
+    private async sendRequest(
+        name: string,
+        url: string,
+        method: string,
+        token: string,
+        eventStream: EventStream,
+        reqObj: any = {},
+        acceptedStatusCode: number[] = [200],
+        headers: any = {},
+    ): Promise<any> {
+        method = method || 'GET';
+        headers[OP_BUILD_USER_TOKEN_HEADER_NAME] = token;
+
+        eventStream.post(new APICallStarted(name, url));
+        const promise = new Promise((resolve, reject) => {
+            request({
+                url,
+                method,
+                headers,
+                json: true,
+                body: reqObj,
+            }, (error, response, respObj) => {
+                if (error) {
+                    eventStream.post(new APICallFailed(name, url, error.message));
+                    reject();
+                }
+
+                if (!acceptedStatusCode.includes(response.statusCode)) {
+                    if (response.statusCode === 401) {

```

Comment:
>statusCode [](start = 33, length = 10)

If status code is 404, throw friendly error message like L21.

Reply:
the information cannot be report precisely， I cannot identify the `URL not existed` or `not provisioned`

---
In reply to: [365630719](https://github.com/docascode/vscode-docs-build/pull/13#discussion_r365630719) [](ancestors = 365630719)

Reply:
Then let API make some change to report error message for 404 case.

---
In reply to: [365639337](https://github.com/docascode/vscode-docs-build/pull/13#discussion_r365639337) [](ancestors = 365639337,365630719)